va/caa.go

```diff
@@ -116,7 +117,61 @@ func parseResults(results []caaResult) (*CAASet, error) {
 	return nil, nil
 }
 
-func (va *ValidationAuthorityImpl) parallelCAALookup(ctx context.Context, name string, lookuper func(context.Context, string) ([]*dns.CAA, error)) []caaResult {
+// If the fqdn has a direct alias among the list of CNAMEs, return it; otherwise
+// return "". Note that RFC 6844 only asks for tree-climbing on direct aliases,
+// not recursive aliases.
+func findAlias(source string, cnames []*dns.CNAME) string {
+	for _, cname := range cnames {
+		if cname.Header().Name == source {
+			return cname.Target
+		}
+	}
+	return ""
+}
+
+func parent(fqdn string) string {
+	labels := strings.Split(fqdn, ".")
+	if len(labels) > 1 {
+		return strings.Join(labels[1:], ".")
+	} else {
+		return ""
+	}
+}
+
+// Implement pre-erratum 5065 style tree-climbing CAA. Note: a strict
+// interpretation of pre-5065 indicates a linear lookup path - if there is any
+// CNAME at all, that precludes further tree-climbing on the original FQDN. This
+// is clearly wrong. We implement a hybrid approach that is strictly more
+// conservative: We always do full tree-climbing on the original FQDN (by virtue
+// of parallelCAALookup. When the LegacyCAA flag is enabled, we also
+// do linear tree climbing on single-level aliases.
+func (va *ValidationAuthorityImpl) treeClimbingLookupCAA(ctx context.Context, fqdn string) ([]*dns.CAA, []*dns.CNAME, error) {
+	target := fqdn
+	// Limit CNAME chasing to break CNAME loops
+	for i := 0; i < 8 && target != ""; i++ {
```

Comment:
Is `8` an arbitrary limit? Should it be a const with a better name?

Reply:
Yes, arbitrary. And updated to a variable with a name (and bumped to 15).

Reply:
I'm not sure it's still enough, please see https://github.com/letsencrypt/boulder/issues/3093